src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/EnvelopedCms.cs

```diff
@@ -0,0 +1,268 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Security.Cryptography.X509Certificates;
+
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography.Pkcs
+{
+    public sealed class EnvelopedCms
+    {
+        //
+        // Constructors
+        //
+
+        public EnvelopedCms()
+            : this(new ContentInfo(Array.Empty<byte>()))
+        {
+        }
+
+        public EnvelopedCms(ContentInfo contentInfo)
+            : this(contentInfo, new AlgorithmIdentifier(Oid.FromOidValue(Oids.TripleDesCbc, OidGroup.EncryptionAlgorithm)))
+        {
+        }
+
+        public EnvelopedCms(ContentInfo contentInfo, AlgorithmIdentifier encryptionAlgorithm)
+        {
+            if (contentInfo == null)
+                throw new ArgumentNullException(nameof(contentInfo));
+            if (encryptionAlgorithm == null)
+                throw new ArgumentNullException(nameof(encryptionAlgorithm));
+
+            Version = 0;  // It makes little sense to ask for a version before you've decoded, but since the desktop returns 0 in that case, we will too.
+            ContentInfo = contentInfo;
+            ContentEncryptionAlgorithm = encryptionAlgorithm;
+            Certificates = new X509Certificate2Collection();
+            UnprotectedAttributes = new CryptographicAttributeObjectCollection();
+            _decryptorPal = null;
+            _lastCall = LastCall.Ctor;
+        }
+
+        //
+        // Latched properties.
+        //     - For senders, the caller sets their values and they act as implicit inputs to the Encrypt() method.
+        //     - For recipients, the Decode() and Decrypt() methods reset their values and the caller can inspect them if desired.
+        //
+
+        public int Version { get; private set; }
+        public ContentInfo ContentInfo { get; private set; }
+        public AlgorithmIdentifier ContentEncryptionAlgorithm { get; private set; }
+        public X509Certificate2Collection Certificates { get; private set; }
+        public CryptographicAttributeObjectCollection UnprotectedAttributes { get; private set; }
+
+        //
+        // Recipients invoke this property to retrieve the recipient information after a Decode() operation. 
+        // Senders should not invoke this property.
+        //
+        public RecipientInfoCollection RecipientInfos
+        {
+            get
+            {
+                switch (_lastCall)
+                {
+                    case LastCall.Ctor:
+                        return new RecipientInfoCollection();
+
+                    case LastCall.Encrypt:
+                        throw PkcsPal.Instance.CreateRecipientInfosAfterEncryptException();
```

Comment:
That error code comes from CryptMsgGetParam(). Apparently that’s its way of saying “Why do you need me to read out the ‘To Line’ on a CMS that you authored?”